### PR TITLE
Feat(cuda): implement cuda circuit bootstrap

### PR DESCRIPTION
**Pull request by bbarbakadze**
#### Add circuit bootstrap in the cuda backend
- Add FP-Keyswitch.
- Add entry points for cuda fk ksk in the public API.
- Add test for fp_ksk in cuda backend.
- Add fixture for bit extract
### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by bbarbakadze** [comment]
@slab-ci gpu_test

**Comment by bbarbakadze** [comment]
@slab-ci gpu_test

**Comment by bbarbakadze** [comment]
@slab-ci gpu_test

File: concrete-core-test/src/lib.rs [hunk 1]
```diff
@@ -7,10 +7,10 @@
 use concrete_core_fixture::{Repetitions, SampleSize};
 
 /// The number of time a test is repeated for a single set of parameter.
-pub const REPETITIONS: Repetitions = Repetitions(10);
+pub const REPETITIONS: Repetitions = Repetitions(1);
```
Comment (agnesLeroy) [inline concrete-core-test/src/lib.rs:10]: You can drop the changes to this file, I merged another change for that already.

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/lwe_ciphertext_discarding_circuit_bootstrap_boolean.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+//use crate::backends::cuda::implementation::engines::{check_base_log, check_glwe_dim};
```
Comment (agnesLeroy) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/lwe_ciphertext_discarding_circuit_bootstrap_boolean.rs:1]: You can remove this file.

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 3]
```diff
@@ -246,6 +246,15 @@ pub fn circuit_bootstrap_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
     let fpksk_input_lwe_key_dimension = fpksk_list.input_lwe_key_dimension();
     let fourier_bsk_output_lwe_dimension = fourier_bsk.output_lwe_dimension();
 
+    println!(
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:249]: I think thoses prints shouldn't be added.

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 4]
```diff
@@ -257,6 +266,15 @@ pub fn circuit_bootstrap_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
     let fpksk_output_polynomial_size = fpksk_list.output_polynomial_size();
     let fpksk_output_glwe_key_dimension = fpksk_list.output_glwe_key_dimension();
 
+    println!(
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:269]: Same here

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 5]
```diff
@@ -309,6 +327,7 @@ pub fn circuit_bootstrap_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
         for pfksk in fpksk_list.fpksk_iter() {
             let mut glwe_out = out_pfksk_buffer_iter.next().unwrap();
             pfksk.private_functional_keyswitch_ciphertext(&mut glwe_out, &lwe_out_bs_buffer);
+            println!("r_glwe_out: {:?}", glwe_out);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:330]: Here again I think you can remove this print

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 6]
```diff
@@ -347,6 +366,7 @@ pub fn homomorphic_shift_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
     stack: DynStack<'_>,
 ) {
     let ciphertext_n_bits = Scalar::BITS;
+    println!("ciphertext_n_bits: {:?}", ciphertext_n_bits);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:369]: Again here

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 7]
```diff
@@ -360,6 +380,9 @@ pub fn homomorphic_shift_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
         &Cleartext(Scalar::ONE << (ciphertext_n_bits - delta_log.0 - 1)),
     );
 
+    //println!("lwe_in: {:?}", lwe_in);
+    //println!("lwe_left_shift_buffer: {:?}", lwe_left_shift_buffer);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:384]: I think you can remove those commented prints

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 8]
```diff
@@ -371,6 +394,8 @@ pub fn homomorphic_shift_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
         CACHELINE_ALIGN,
         |_| Scalar::ZERO,
     );
+
+    //println!("pbs_accumulator_data: {:?}", pbs_accumulator_data);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:398]: Same here

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 9]
```diff
@@ -384,6 +409,8 @@ pub fn homomorphic_shift_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
             Scalar::ONE << (ciphertext_n_bits - 1 - base_log_cbs.0 * level_count_cbs.0),
         ));
 
+    //println!("pbs_accumulator: {:?}", pbs_accumulator.get_body().tensor);
+    //println!("lwe_left_shift_buffer: {:?}", lwe_left_shift_buffer);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:413]: Same here

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs [hunk 10]
```diff
@@ -394,6 +421,16 @@ pub fn homomorphic_shift_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
...
+    println!("r_lwe_in: {:?}", lwe_left_shift_buffer.get_body_and_mask());
+    println!(
+        "r_pbs_accumulator_mask: {:?}",
+        pbs_accumulator.get_mask().tensor
+    );
+    println!(
+        "r_pbs_accumulator_body: {:?}",
+        pbs_accumulator.get_body().tensor
+    );
+    println!("r_pbs_lwe_out: {:?}", lwe_out);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/mod.rs:433]: Same here

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/tests.rs [hunk 11]
```diff
@@ -181,7 +181,7 @@ pub fn test_extract_bits() {
 fn test_circuit_bootstrapping_binary() {
     // Define settings for an insecure toy example
     let polynomial_size = PolynomialSize(512);
-    let glwe_dimension = GlweDimension(2);
+    let glwe_dimension = GlweDimension(1);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/tests.rs:184]: I think we should put back the parameters to what they were

File: concrete-core/src/backends/fft/private/crypto/wop_pbs/tests.rs [hunk 12]
```diff
@@ -190,7 +190,7 @@ fn test_circuit_bootstrapping_binary() {
     let level_pksk = DecompositionLevelCount(2);
     let base_log_pksk = DecompositionBaseLog(15);
 
-    let level_count_cbs = DecompositionLevelCount(1);
+    let level_count_cbs = DecompositionLevelCount(2);
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fft/private/crypto/wop_pbs/tests.rs:193]: Same here

File: concrete-core/src/commons/crypto/glwe/keyswitch.rs [hunk 13]
```diff
@@ -1403,6 +1403,8 @@ impl<Cont> LwePrivateFunctionalPackingKeyswitchKey<Cont> {
         ck_dim_eq!(self.input_lwe_key_dimension().0  => before.lwe_size().to_lwe_dimension().0 );
         ck_dim_eq!(self.output_glwe_key_dimension().0 => after.size().to_glwe_dimension().0);
 
+        println!("----------------FP-KSK-------------------------------------");
+        //println!("before: {:?}", before.get_body_and_mask());
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/glwe/keyswitch.rs:1407]: I think you can remove those lines

File: concrete-core/src/commons/crypto/glwe/keyswitch.rs [hunk 14]
```diff
@@ -1414,6 +1416,10 @@ impl<Cont> LwePrivateFunctionalPackingKeyswitchKey<Cont> {
             let rounded = decomposer.closest_representable(*input_lwe);
             let decomp = decomposer.decompose(rounded);
 
+            // println!("input_lwe: {:?}", *input_lwe);
+            println!("input_lwe_rounded: {:?}", rounded);
+
+            //println!("block: {:?}", block);
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/glwe/keyswitch.rs:1422]: Same here

File: concrete-core/src/commons/crypto/glwe/keyswitch.rs [hunk 15]
```diff
@@ -1423,6 +1429,8 @@ impl<Cont> LwePrivateFunctionalPackingKeyswitchKey<Cont> {
                 .rev()
                 .zip(decomp)
             {
+                //println!("level_key_cipher:{:?}", level_key_cipher);
+                //println!("decomposed:{:?}", decomposed);
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/glwe/keyswitch.rs:1433]: Same here

File: concrete-cuda/cuda/src/keyswitch.cuh [hunk 16]
```diff
@@ -17,6 +17,113 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
...
+      Torus decomposed = decompose_one<Torus>(state, mod_b_mask, base_log);
+      local_glwe_chunk[tid] -= decomposed * ksk_glwe_chunk[tid];
+    }
+  }
+
+  cur_glwe_chunk[tid] = local_glwe_chunk[tid];
+
+  __syncthreads();
+  if (threadIdx.x == 0 && blockIdx.x == 0 && blockIdx.y == 0) {
+    printf("d_out_glwe: ");
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/keyswitch.cuh:117]: I think you can remove these prints

File: concrete-cuda/cuda/src/keyswitch.cuh [hunk 17]
```diff
@@ -17,6 +17,113 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
...
+
+
+  // iterate through each coefficient of  input lwe
+  for (size_t i = 0; i <= lwe_dimension_in; i++) {
+    Torus a_i =
+        round_to_closest_multiple(cur_input_lwe[i], base_log, level_count);
+
+    __syncthreads();
+    if (threadIdx.x == 0 && blockIdx.x == 0 && blockIdx.y == 0) {
+        printf("rounded: %lu\n", a_i);
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/keyswitch.cuh:94]: I think you can remove this print

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 18]
```diff
@@ -683,4 +789,84 @@ void host_blind_rotate_and_sample_extraction(
...
+    uint32_t base_log_bsk, uint32_t level_pksk, uint32_t base_log_pksk,
+    uint32_t level_cbs, uint32_t base_log_cbs, uint32_t number_of_samples,
+    uint32_t max_shared_memory)
+{
+  auto stream = static_cast<cudaStream_t *>(v_stream);
+
+  uint32_t ciphertext_n_bits = sizeof(Torus) * 8;
+  uint32_t lwe_size = lwe_dimension + 1;
+  int pbs_count = number_of_samples * level_cbs;
+  printf("circuit bootstrap #1\n");
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/bootstrap_wop.cuh:811]: I think you can remove this print

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 19]
```diff
@@ -426,6 +473,65 @@ __global__ void fill_lut_body_for_current_bit(Torus *lut, Torus value) {
...
+    tid += params::degree / params::opt;
+  }
+
+
+  __syncthreads();
+  if (blockIdx.x == 0 && threadIdx.x == 0 && 0) {
+    for (int k = 0; k < gridDim.x; k++) {
+      auto cur_prt = &lut[k * 2 * params::degree + params::degree];
+      for (int i = 0; i < params::degree; i++) {
+        printf("d_lwe_shift_%d[%d]: %lu\n", k, i, cur_prt[i]);
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/bootstrap_wop.cuh:500]: I think you can remove this print

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 20]
```diff
@@ -377,6 +378,52 @@ __global__ void copy_and_shift_lwe(Torus *dst_copy, Torus *dst_shift,
...
+  size_t opt = lwe_size / threads_per_block;
+  size_t rem = lwe_size & (threads_per_block - 1);
+
+  auto cur_dst = &dst_shift[blockId * lwe_size];
+  auto cur_src = &src[blockIdx.y * lwe_size];
+
+  __syncthreads();
+  if (blockIdx.x == 0 && threadIdx.x == 0 && blockIdx.y == 0 && 0) {
+    for (int i = 0; i < lwe_size * gridDim.x * gridDim.y; i++) {
+      printf("d_lwe_in_%d[%d]: %lu\n", i / lwe_size,  i % lwe_size, src[i]);
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/bootstrap_wop.cuh:399]: I think you can remove this print

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 21]
```diff
@@ -377,6 +378,52 @@ __global__ void copy_and_shift_lwe(Torus *dst_copy, Torus *dst_shift,
...
+    tid += threads_per_block;
+  }
+
+  if (threadIdx.x < rem)
+    cur_dst[tid] = cur_src[tid] * value;
+
+  __syncthreads();
+  if (blockIdx.x == 0 && threadIdx.x == 0 && blockIdx.y == 0 && 0) {
+    for (int i = 0; i < lwe_size * gridDim.x * gridDim.y; i++) {
+      printf("d_lwe_shift[%d]: %lu\n",  (i % lwe_size),
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/bootstrap_wop.cuh:417]: I think you can remove this print

File: concrete-cuda/cuda/src/keyswitch.cuh [hunk 22]
```diff
@@ -17,6 +17,113 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
...
+  size_t chunk_id = blockIdx.x;
+  size_t ksk_id = ciphertext_id % number_of_keys;
+  __syncthreads();
+  if (threadIdx.x == 0 && blockIdx.x == 0 && blockIdx.y == 0) {
+    printf("glwe_size: %lu\n", glwe_size);
+    printf("lwe_size: %lu\n", lwe_size);
+    printf("ksk_size: %lu\n", ksk_size);
+    printf("ksk_block_size: %lu\n", ksk_block_size);
+    printf("coef_per_block: %lu\n", coef_per_block);
+    printf("number_of_keys: %lu\n", number_of_keys);
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/keyswitch.cuh:51]: I think you can remove those prints

File: concrete-cuda/cuda/src/keyswitch.cuh [hunk 23]
```diff
@@ -17,6 +17,113 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
...
+  local_glwe_chunk[tid] = 0;
+
+  __syncthreads();
+  if (threadIdx.x == 0 && blockIdx.x == 0 && blockIdx.y == 0) {
+    Torus g = 0;
+    for (size_t i = 0; i < ksk_size; i++) {
+      g += cur_ksk[i];
+    }
+
+    printf("g = %lu \n", g);
```
Comment (agnesLeroy) [inline concrete-cuda/cuda/src/keyswitch.cuh:82]: This one as well

**Review by agnesLeroy — CHANGES_REQUESTED** [review]
Hi @bbarbakadze! Here comes a first review :slightly_smiling_face:

**Review by agnesLeroy — APPROVED** [review]

**Comment by bbarbakadze** [comment]
@slab-ci gpu_test

**Comment by bbarbakadze** [comment]
@slab-ci gpu_test

**Pull request merged**